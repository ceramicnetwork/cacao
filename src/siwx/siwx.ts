import type { Cacao } from '../cacao.js'
import { ParsedMessage as ABNFParsedMessage } from '../abnf.js'
import { AccountId, ChainId } from 'caip'

/**
 * Possible message error types.
 */
export enum ErrorTypes {
  /**Thrown when the `validate()` function can verify the message. */
  INVALID_SIGNATURE = 'Invalid signature.',
  /**Thrown when the `expirationTime` is present and in the past. */
  EXPIRED_MESSAGE = 'Expired message.',
  /**Thrown when some required field is missing. */
  MALFORMED_SESSION = 'Malformed session.',
}

/**
 * Possible signature types that this library supports.
 */
export enum SignatureType {
  /**EIP-191 signature scheme */
  PERSONAL_SIGNATURE = 'Personal signature',
}

export class SiwxMessage {
  /**RFC 4501 dns authority that is requesting the signing. */
  domain: string
  /**Ethereum address performing the signing conformant to capitalization
   * encoded checksum specified in EIP-55 where applicable. */
  address: string
  /**Human-readable ASCII assertion that the user will sign, and it must not
   * contain `\n`. */
  statement?: string
  /**RFC 3986 URI referring to the resource that is the subject of the signing
   *  (as in the __subject__ of a claim). */
  uri: string
  /**Current version of the message. */
  version: string
  /**Randomized token used to prevent replay attacks, at least 8 alphanumeric
   * characters. */
  nonce?: string
  /**ISO 8601 datetime string of the current time. */
  issuedAt?: string
  /**ISO 8601 datetime string that, if present, indicates when the signed
   * authentication message is no longer valid. */
  expirationTime?: string
  /**ISO 8601 datetime string that, if present, indicates when the signed
   * authentication message will become valid. */
  notBefore?: string
  /**System-specific identifier that may be used to uniquely refer to the
   * sign-in request. */
  requestId?: string
  /**EIP-155 Chain ID to which the session is bound, and the network where
   * Contract Accounts must be resolved. */
  chainId: string
  /**List of information or references to information the user wishes to have
   * resolved as part of authentication by the relying party. They are
   * expressed as RFC 3986 URIs separated by `\n- `. */
  resources?: Array<string>
  /**Signature of the message signed by the wallet. */
  signature?: string
  /**Type of sign message to be generated. */
  type?: SignatureType

  constructor(param: string | Partial<SiwxMessage>) {
    if (typeof param === 'string') {
      const parsedMessage = new ABNFParsedMessage(param)
      this.domain = parsedMessage.domain
      this.address = parsedMessage.address
      this.statement = parsedMessage.statement
      this.uri = parsedMessage.uri
      this.version = parsedMessage.version
      this.nonce = parsedMessage.nonce
      this.issuedAt = parsedMessage.issuedAt
      this.expirationTime = parsedMessage.expirationTime
      this.notBefore = parsedMessage.notBefore
      this.requestId = parsedMessage.requestId
      this.chainId = parsedMessage.chainId
      this.resources = parsedMessage.resources
    } else {
      Object.assign(this, param)
    }
    if (!this.chainId) {
      throw new Error(`Chain ID is mandatory`)
    }
  }

  static fromCacao<T extends SiwxMessage>(this: new (...args: Array<any>) => T, cacao: Cacao): T {
    const account = AccountId.parse(cacao.p.iss.replace('did:pkh:', ''))
    const siwx = new this({
      domain: cacao.p.domain,
      address: account.address,
      uri: cacao.p.aud,
      version: cacao.p.version,
      chainId: new ChainId(account.chainId).reference,
    })

    if (cacao.p.statement) siwx.statement = cacao.p.statement
    if (cacao.p.nonce) siwx.nonce = cacao.p.nonce
    if (cacao.p.iat) siwx.issuedAt = cacao.p.iat
    if (cacao.p.exp) siwx.expirationTime = cacao.p.exp
    if (cacao.p.nbf) siwx.notBefore = cacao.p.nbf
    if (cacao.p.requestId) siwx.requestId = cacao.p.requestId
    if (cacao.p.resources) siwx.resources = cacao.p.resources

    if (cacao.s) {
      if (cacao.s.s) siwx.signature = cacao.s.s
      if (cacao.s.t === 'eip191') siwx.type = SignatureType.PERSONAL_SIGNATURE
    }

    return siwx
  }

  toMessage(chain: string): string {
    const header = `${this.domain} wants you to sign in with your ${chain} account:`
    const uriField = `URI: ${this.uri}`
    let prefix = [header, this.address].join('\n')
    const versionField = `Version: ${this.version}`
    const chainIdField = `Chain ID: ${this.chainId}`

    if (!this.nonce) {
      this.nonce = (Math.random() + 1).toString(36).substring(4)
    }

    const nonceField = `Nonce: ${this.nonce}`

    const suffixArray = [uriField, versionField, chainIdField, nonceField]

    if (this.issuedAt) {
      Date.parse(this.issuedAt)
    }
    this.issuedAt = this.issuedAt ? this.issuedAt : new Date().toISOString()
    suffixArray.push(`Issued At: ${this.issuedAt}`)

    if (this.expirationTime) {
      const expiryField = `Expiration Time: ${this.expirationTime}`

      suffixArray.push(expiryField)
    }

    if (this.notBefore) {
      suffixArray.push(`Not Before: ${this.notBefore}`)
    }

    if (this.requestId) {
      suffixArray.push(`Request ID: ${this.requestId}`)
    }

    if (this.resources) {
      suffixArray.push([`Resources:`, ...this.resources.map((x) => `- ${x}`)].join('\n'))
    }

    const suffix = suffixArray.join('\n')

    if (this.statement) {
      prefix = [prefix, this.statement].join('\n\n')
    }

    return [prefix, suffix].join('\n\n')
  }
}
