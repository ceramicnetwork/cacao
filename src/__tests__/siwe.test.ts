import { SiweMessage } from '../siwx/siwe.js'
import { SignatureType } from '../siwx/siwx.js'
import { Wallet } from '@ethersproject/wallet'

describe('Message Generation', () => {
  test('With optional fields', () => {
    const msg = new SiweMessage({
      domain: 'service.org',
      address: '0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2',
      statement: 'I accept the ServiceOrg Terms of Service: https://service.org/tos',
      uri: 'https://service.org/login',
      version: '1',
      nonce: '32891757',
      issuedAt: '2021-09-30T16:25:24.000Z',
      chainId: '1',
      resources: [
        'ipfs://Qme7ss3ARVgxv6rXqVPiikMJ8u2NLgmgszg13pYrDKEoiu',
        'https://example.com/my-web2-claim.json',
      ],
    })

    expect(msg.toMessage()).toEqual(
      'service.org wants you to sign in with your Ethereum account:\n0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2\n\nI accept the ServiceOrg Terms of Service: https://service.org/tos\n\nURI: https://service.org/login\nVersion: 1\nNonce: 32891757\nIssued At: 2021-09-30T16:25:24.000Z\nChain ID: 1\nResources:\n- ipfs://Qme7ss3ARVgxv6rXqVPiikMJ8u2NLgmgszg13pYrDKEoiu\n- https://example.com/my-web2-claim.json'
    )
  })

  test('No optional field', () => {
    const msg = new SiweMessage({
      domain: 'service.org',
      address: '0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2',
      statement: 'I accept the ServiceOrg Terms of Service: https://service.org/tos',
      uri: 'https://service.org/login',
      version: '1',
      nonce: '32891757',
      issuedAt: '2021-09-30T16:25:24.000Z',
    })

    expect(msg.toMessage()).toEqual(
      'service.org wants you to sign in with your Ethereum account:\n0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2\n\nI accept the ServiceOrg Terms of Service: https://service.org/tos\n\nURI: https://service.org/login\nVersion: 1\nNonce: 32891757\nIssued At: 2021-09-30T16:25:24.000Z'
    )
  })

  test('Timestamp without microseconds', () => {
    const msg = new SiweMessage({
      domain: 'service.org',
      address: '0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2',
      statement: 'I accept the ServiceOrg Terms of Service: https://service.org/tos',
      uri: 'https://service.org/login',
      version: '1',
      nonce: '32891757',
      issuedAt: '2021-09-30T16:25:24Z',
    })

    expect(msg.toMessage()).toEqual(
      'service.org wants you to sign in with your Ethereum account:\n0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2\n\nI accept the ServiceOrg Terms of Service: https://service.org/tos\n\nURI: https://service.org/login\nVersion: 1\nNonce: 32891757\nIssued At: 2021-09-30T16:25:24Z'
    )
  })
})

describe('Ocap', () => {
  const wallet = Wallet.fromMnemonic(
    'despair voyage estate pizza main slice acquire mesh polar short desk lyrics'
  )

  test('Sign message using SiweMessage object', async () => {
    const address = wallet.address
    const fixedDate = new Date('2021-10-14T07:18:41Z')
    const msg = new SiweMessage({
      domain: 'self.id',
      statement: 'Give this app access to your streams',
      nonce: '12345678',
      type: SignatureType.PERSONAL_SIGNATURE,
      address: address,
      chainId: '1',
      issuedAt: fixedDate.toISOString(),
      resources: ['ceramic://bagcqcerakszw2vsovxznyp5gfnpdj4cqm2xiv76yd24wkjewhhykovorwo6a'],
      uri: `did:pkh:eip155:1:${address}`,
      version: '0.1',
    })
    expect(msg.toMessage()).toMatchSnapshot()

    const signature = await wallet.signMessage(msg.toMessage())
    expect(signature).toMatchSnapshot()
  })
})
