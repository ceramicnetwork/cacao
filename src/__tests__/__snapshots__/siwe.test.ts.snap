// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Ocap Sign message using SiweMessage object 1`] = `
"self.id wants you to sign in with your Ethereum account:
0xBd9D9c7DC389715a89fC8149E4a5Be91336B2796

Give this app access to your streams

URI: did:pkh:eip155:1:0xBd9D9c7DC389715a89fC8149E4a5Be91336B2796
Version: 0.1
Chain ID: 1
Nonce: 12345678
Issued At: 2021-10-14T07:18:41.000Z
Resources:
- ceramic://bagcqcerakszw2vsovxznyp5gfnpdj4cqm2xiv76yd24wkjewhhykovorwo6a"
`;

exports[`Ocap Sign message using SiweMessage object 2`] = `"0x66e3a8d71a0c15bbec0afe1e6805cb2826375a6d160031364e3d780aecd9fd0630d48860c3a816b51ca108bc8d6545cb8ae900fa9f988cb3fce1de333d31a43d1b"`;
